### PR TITLE
feat: add prTitleStrict config option to bypass prTitle manipulation

### PR DESCRIPTION
**Pull request by setchy**
## Changes

Adds a new optional boolean config, `prTitleStrict`, which is defaulted to `false`.

## Context

This new config option allows users to bypass the prTitle manipulation logic which will
- append the `baseBranch` if one is set
- append the `updateType` if it is a grouped update and `separateMajorMinor` or `separateMinorPatch` is true

Tested against: https://github.com/setchy/renovate-demo-split-major-suffix

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — CHANGES_REQUESTED** [review]
🤔 those are not applied when not using the preset. we need to find a better way

**Comment by setchy** [comment]
> 🤔 those are not applied when not using the preset. we need to find a better way

gotcha.  

so to confirm, find an alternate way that works for all scenarios where `separateMajorMinor:true` and `separateMinorPatch:true`.  Is that right?

could that be a migration rule (I haven't spent much time looking into how these work to-date)?

**Comment by viceice** [comment]
can probably work with migration, but will break any other `commitMessageSuffix` usage. 🤔

btw: PR title seems wrong, prefix vs suffix 🙃

**Comment by setchy** [comment]
> btw: PR title seems wrong, prefix vs suffix 🙃

🤦 🙃 😓

**Comment by setchy** [comment]
Maybe to take a step back for a minute and confirm the assumption I've been making which is existing title structures should be maintained.

How important @viceice is it to keep the `prTitle` manipulation logic which appends ether `(major)`, `(minor)` or `(patch)` if the PR is a group and one of the seperateMajorMinor or separateMinorPatch flags is configured?  

Does that behavior need to be maintained?  If it was removed, there would be some amount of PRs autoclosed and reopened, but otherwise everything else would continue to function as-is.  And if someone did want to keep it, then they could set `commitMessageSuffix: "({{updateType}})` as they see fit.

**Comment by setchy** [comment]
For further context, my motivation behind this change is to be able to use `commitMessageSuffix`, as the OSV `[SECURITY]` PRs do, to suffix all PRs with their updateType for metrics, analysis and assigning important/priority.  

This has been working well for us, with the exception of the above scenarios which results in PRs named `..... (major) [MAJOR]` etc

**Comment by setchy** [comment]
alternatively, @viceice - I could add a boolean config flag that defaults to `true`, which when set to `false` skips the prTitle manipulation logic.  This would avoid it being a `breaking change`

**Comment by viceice** [comment]
sounds interesting
@rarkins WDYT?

**Comment by rarkins** [comment]
I'm in favor of removing this hardcoded PR title logic and making it able to be overridden. 

I don't think that doing this via presets is a good idea though, because many if not most people would configure settings like `separateMultipleMajor` directly.

We particularly worry about PR titles changing when it might mean reopening a lot of closed/ignored PRs.

In many cases group PRs are "immortal" anyway, so they can't be in a closed/ignored state. If they were open, then changing title as a result of this PR would simply mean the PR is renamed, which I'm fine with.

However there would be some like "Update aws packages to v3 (major)` which would be affected if the title changed.

An option as proposed to simply skip the current behavior could be ok, perhaps naming it is the hardest part.

**Comment by setchy** [comment]
> perhaps naming it is the hardest part.

Absolutely 😅 


Two follow up questions ...

Should this new `skip PR title appending` also include the `({{baseBranch}})` piece
https://github.com/renovatebot/renovate/blob/3d19f0a64887c3b9f9cdf3823170389d3d35be20/lib/workers/repository/updates/generate.ts#L258

Also, more out of curiosity, why the triple template evaluations one after each-other?
https://github.com/renovatebot/renovate/blob/3d19f0a64887c3b9f9cdf3823170389d3d35be20/lib/workers/repository/updates/generate.ts#L238-L243

**Comment by setchy** [comment]
Pivoted to a non-breaking boolean config option.  The name is a mouthful - alternate suggestions welcomed 🙏

**Comment by rarkins** [comment]
If we combine this with the baseBranch part then it could be called prTitleStrict and default to false. If true, do not manipulate the prTitle for base branch or update type? And we try to move them to templates later

**Comment by setchy** [comment]
> If we combine this with the baseBranch part then it could be called prTitleStrict and default to false. If true, do not manipulate the prTitle for base branch or update type? And we try to move them to templates later

Much more elegant name.  Updated

**Review by viceice — COMMENTED** [review]
I'm still not happy with that.

@rarkins ideas?

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
@viceice unhappy in what way?

File: lib/config/options/index.ts [hunk 1]
```diff
@@ -1855,6 +1855,14 @@ const options: RenovateOptions[] = [
     default: null,
     cli: false,
   },
+  {
+    name: 'prTitleStrict',
+    description:
+      'Whether to bypass appending to the Pull Request title the `baseBranch` or `updateType` if update is part of a group and either `separateMajorMinor` or `separateMinorPatch` is true.',
```
[the remark this answers is not in the document]
Comment (setchy) [reply lib/config/options/index.ts]: Good suggestion @rarkins - updated

**Comment by setchy** [comment]
@viceice @rarkins - any further feedback 🙏

File: lib/config/options/index.ts [hunk 2]
```diff
@@ -1855,6 +1855,14 @@ const options: RenovateOptions[] = [
     default: null,
     cli: false,
   },
+  {
+    name: 'prTitleStrict',
+    description:
+      'Whether to bypass appending extra context to the Pull Request title.',
+    type: 'boolean',
```
Comment (rarkins) [inline lib/config/options/index.ts:1862]:
```suggestion
    type: 'boolean',
    experimental: true,
```

Comment (rarkins) [reply lib/config/options/index.ts:1862]: Maybe let's mark it as experimental and move forward?

Comment (setchy) [reply lib/config/options/index.ts:1862]: Sounds good to me - thanks @rarkins

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**